diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/AvoidNestedFunctionsRule.kt

```diff
@@ -0,0 +1,98 @@
+import org.cqfn.diktat.ruleset.utils.findAllNodesWithSpecificType
+import org.cqfn.diktat.ruleset.utils.getFirstChildWithType
+import org.cqfn.diktat.ruleset.utils.hasChildOfType
+import org.cqfn.diktat.ruleset.utils.hasParent
+import org.jetbrains.kotlin.com.intellij.lang.ASTNode
+import org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.PsiWhiteSpaceImpl
+import org.jetbrains.kotlin.com.intellij.psi.tree.IElementType
+import org.jetbrains.kotlin.psi.KtFunction
+import org.jetbrains.kotlin.psi.psiUtil.parents
+
+/**
+ * This rule checks for nested functions and warns if it finds any.
+ */
+class AvoidNestedFunctionsRule(private val configRules: List<RulesConfig>) : Rule("avoid-nested-functions") {
+    private lateinit var emitWarn: ((offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit)
+    private var isFixMode: Boolean = false
+
+    override fun visit(node: ASTNode, autoCorrect: Boolean, emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit) {
+        emitWarn = emit
+        isFixMode = autoCorrect
+
+        if (node.elementType == FUN) {
+            handleNestedFunctions(node)
+        }
+    }
+
+    // FixMe: need to detect all properties, which local function is using and add them to params of this function
+    @Suppress("UnsafeCallOnNullableType")
+    private fun handleNestedFunctions(node: ASTNode) {
+        if (isNestedFunction(node)) {
+            val funcName = node.getFirstChildWithType(IDENTIFIER)!!.text
+
+            AVOID_NESTED_FUNCTIONS.warnAndFix(configRules, emitWarn, isFixMode, "fun $funcName", node.startOffset, node,
+                    canBeAutoCorrected = checkFunctionReferences(node)) {
+                // We take last nested function, then add and remove child from bottom to top
+                val lastFunc = node.findAllNodesWithSpecificType(FUN).last()
+                val funcSeq = lastFunc.parents().filter { it.elementType == FUN }.toMutableList()
+                funcSeq.add(0, lastFunc)
+                val firstFunc = funcSeq.last()
+                funcSeq.dropLast(1).forEachIndexed { index, it ->
+                    val parent = funcSeq[index + 1]
+                    if (it.treePrev.isWhiteSpaceWithNewline()) {
+                        parent.removeChild(it.treePrev)
+                    }
+                    firstFunc.treeParent.addChild(it.clone() as ASTNode, firstFunc)
+                    firstFunc.treeParent.addChild(PsiWhiteSpaceImpl("\n"), firstFunc)
+                    parent.removeChild(it)
+                }
+            }
+        }
+    }
+
+    private fun isNestedFunction(node: ASTNode): Boolean =
+            node.hasParent(FUN) && node.hasFunParentUntil(CLASS_BODY) && !node.hasChildOfType(MODIFIER_LIST)
+
+
+    private fun ASTNode.hasFunParentUntil(stopNode: IElementType): Boolean =
+            parents().asSequence().toList().takeWhile { it.elementType != stopNode }.any { it.elementType == FUN }
```

Comment:
```suggestion
            parents().asSequence().takeWhile { it.elementType != stopNode }.any { it.elementType == FUN }
```
Why do you need `toList` here?